controllers/application_controller.go

```diff
@@ -0,0 +1,350 @@
+	err = r.syncApplicationStatus(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to sync application status")
+		return err
+	}
+	return nil
+}
+
+func (r *ApplicationReconciler) fixProject(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) (removed bool, err error) {
+	logger := log.FromContext(ctx)
+
+	ns := &corev1.Namespace{}
+	err = r.client.Get(ctx, client.ObjectKey{Name: tenantApp.GetNamespace()}, ns)
+	if err != nil {
+		return
+	}
+	tenantName := ns.Labels[constants.OwnerTenant]
+	if tenantName == "" {
+		if argocdApp != nil && argocdApp.GetDeletionTimestamp() == nil {
+			// Applications on namespaces that do not belong to a tenant will be removed.
+			logger.Info("Remove unmanaged application")
+			err = r.client.Delete(ctx, argocdApp)
+			if err != nil {
+				r.recorder.Eventf(tenantApp, corev1.EventTypeWarning, "RemoveApplicationFailed", "Failed to remove unmanaged application", err)
+				return
+			}
+			r.recorder.Eventf(tenantApp, corev1.EventTypeNormal, "ApplicationRemoved", "Remove unmanaged application succeeded")
+		}
+		removed = true
+		return
+	}
+	project, found, err := unstructured.NestedString(tenantApp.UnstructuredContent(), "spec", "project")
+	if err != nil {
+		return
+	}
+	if !found {
+		err = errors.New("spec.project not found")
+		return
+	}
+	if project != tenantName {
+		logger.Info("Overwrite project", "before", project, "after", tenantName)
+		newApp := argocd.Application()
+		newApp.SetNamespace(tenantApp.GetNamespace())
+		newApp.SetName(tenantApp.GetName())
+		err = unstructured.SetNestedField(newApp.UnstructuredContent(), tenantName, "spec", "project")
+		if err != nil {
+			return
+		}
+		err = r.client.Patch(ctx, newApp, client.Apply, &client.PatchOptions{
+			Force:        pointer.BoolPtr(true),
+			FieldManager: constants.ProjectFieldManager,
+		})
+		if err != nil {
+			r.recorder.Eventf(tenantApp, corev1.EventTypeWarning, "FixProjectFailed", "Failed to fix application project", err)
+			return
+		}
+		r.recorder.Eventf(tenantApp, corev1.EventTypeNormal, "ProjectFixed", "Fix application project succeeded")
+		return
+	}
+	return
+}
+
+func (r *ApplicationReconciler) syncApplicationSpec(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) error {
+	logger := log.FromContext(ctx)
+
+	labels := make(map[string]string)
+	for k, v := range tenantApp.GetLabels() {
+		if strings.Contains(k, "kubernetes.io/") {
+			continue
+		}
+		labels[k] = v
+	}
+	labels[constants.OwnerAppNamespace] = tenantApp.GetNamespace()
+
+	annotations := make(map[string]string)
+	for k, v := range tenantApp.GetAnnotations() {
+		if strings.Contains(k, "kubernetes.io/") {
+			continue
+		}
+		annotations[k] = v
+	}
+	var finalizers []string
+	for _, fin := range tenantApp.GetFinalizers() {
+		if fin == argocd.ResourcesFinalizer {
+			finalizers = append(finalizers, fin)
```

Comment:
Is it correct that we apply the Argo CD finalizer?

Reply:
correct.
Tenant users will want to specify the finalizer.